llms/mlx_lm/utils.py

```diff
@@ -91,10 +91,34 @@ def get_model_path(path_or_hf_repo: str) -> Path:
     return model_path
 
 
+def apply_repetition_penalty(logits: mx.array, generated_tokens: Any, penalty: float):
+    """
+    Apply repetition penalty to specific logits based on the given context.
+
+    Args:
+        logits (mx.array): The logits produced by the language model.
+        generated_tokens (any): A list of N previous tokens.
+        penalty (float): The repetition penalty factor to be applied.
+
+    Returns:
+        logits (mx.array): Logits with repetition penalty applied to generated tokens.
+    """
+    if len(generated_tokens) > 0:
+        indices = mx.array([token for token in generated_tokens])
+        selected_logits = logits[:, indices]
+        selected_logits = mx.where(
+            selected_logits < 0, selected_logits * penalty, selected_logits / penalty
+        )
+        logits[:, indices] = selected_logits
+    return logits
+
+
 def generate_step(
     prompt: mx.array,
     model: nn.Module,
-    temp: float = 0.0,
+    temp: 0.0,
+    repetition_penalty: Optional[float] = None,
+    repetition_context_size: Optional[int] = None,
```

Comment:
Is `None` a reasonable default for the `repetition_context_size`? IIRC it will use the full sequence if this is `None`? Maybe we should set it at a more reasonable default? What do you typically use when you run this?

Reply:
Yes if `repetition_context_size` is `none` it will be applied for the entire sequence. I feel that it's a bit tricky to set a default repetition context size as it is subjective. Since we are setting the `max_tokens` to be 100 as default in `generate` I think we can set a default `repetition_context_size` with respect to `max_tokens` being 100? Probably 20 or 25?

@mzbac @awni  any thoughts on this ?

Reply:
Anything in the 10-20 range seems reasonable to me. I would lean towards less penalty as a default rather than higher so using a smaller default repetition context size seems more in that direction.